Sources/StorageCommitmentScpCallback.cpp

```diff
@@ -0,0 +1,187 @@
+  void**              handler /* out */,
+  const char*         jobId,
+  const char*         transactionUid,
+  const char* const*  sopClassUids,
+  const char* const*  sopInstanceUids,
+  uint32_t            countInstances,
+  const char*         remoteAet,
+  const char*         calledAet)
+{
+  try
+  {
+    PythonLock lock;
+
+    PythonObject args(lock, PyTuple_New(6));
+    {
+      PythonString str(lock, jobId);
+      PyTuple_SetItem(args.GetPyObject(), 0, str.Release());
+    }
+    {
+      PythonString str(lock, transactionUid);
+      PyTuple_SetItem(args.GetPyObject(), 1, str.Release());
+    }
+    {
+      PythonObject sopClassUidList(lock, PyList_New(countInstances));
+      for (uint32_t i = 0; i < countInstances; i++)
+      {
+        PythonString str(lock, sopClassUids[i]);
+        PyList_SetItem(sopClassUidList.GetPyObject(), i, str.Release());
+      }
+      PyTuple_SetItem(args.GetPyObject(), 2, sopClassUidList.Release());
+      PythonObject sopInstanceUidList(lock, PyList_New(countInstances));
+      for (uint32_t i = 0; i < countInstances; i++)
+      {
+        PythonString str(lock, sopInstanceUids[i]);
+        PyList_SetItem(sopInstanceUidList.GetPyObject(), i, str.Release());
+      }
+      PyTuple_SetItem(args.GetPyObject(), 3, sopInstanceUidList.Release());
+    }
+    {
+      PythonString str(lock, remoteAet);
+      PyTuple_SetItem(args.GetPyObject(), 4, str.Release());
+    }
+    {
+      PythonString str(lock, calledAet);
+      PyTuple_SetItem(args.GetPyObject(), 5, str.Release());
+    }
+
+    PythonObject result(lock, PyObject_CallObject(storageCommitmentScpCallback_, args.GetPyObject()));
+    *handler = result.Release();
+
+    std::string traceback;
+    if (lock.HasErrorOccurred(traceback))
+    {
+      OrthancPlugins::LogError("Error in the Python storage commitment SCP callback, "
+                               "traceback:\n" + traceback);
+      return OrthancPluginErrorCode_Plugin;
+    }
+  }
+  catch (OrthancPlugins::PluginException& e)
+  {
+    OrthancPlugins::LogError("Error in the Python storage commitment SCP callback: " +
+                             std::string(e.What(OrthancPlugins::GetGlobalContext())));
+  }
+  return OrthancPluginErrorCode_Success;
+}
+
+static OrthancPluginErrorCode StorageCommitmentLookupCallback(
+  OrthancPluginStorageCommitmentFailureReason* target /* out */,
+  void*                                        handler,
+  const char*                                  sopClassUid,
+  const char*                                  sopInstanceUid)
+{
+  try
+  {
+    PythonLock lock;
+
+    PythonObject args(lock, PyTuple_New(3));
+    {
+      PythonString str(lock, sopClassUid);
+      PyTuple_SetItem(args.GetPyObject(), 0, str.Release());
+    }
+    {
+      PythonString str(lock, sopInstanceUid);
+      PyTuple_SetItem(args.GetPyObject(), 1, str.Release());
+    }
+    {
+      PyObject* data = (PyObject*) handler;
```

Comment:
For consistency reason, I would keep the same order for the arguments in C++ and in python: handler, sopClassUid and then sopInstanceUid

Reply:
I intentionally ordered the arguments as `[sopClassUid, sopInstanceUid, handler]` due to the following reasons:
1. The `handler`/`data` argument for Python does not serve the same purpose as `OrthancPlugins::IStorageCommitmentScpHandler` used in C++. (`handler` in C++ is a class that contains a factory method; while `handler`/`data` in Python stores an arbitrary data provided by the user.)
2. The original C++ callback does not contain the `handler`/`data` argument. Since it is a newly introduced argument, I think it should be appended to the argument list.

More specifically, when writing C++ plugins (https://hg.orthanc-server.com/orthanc/file/tip/OrthancServer/Plugins/Samples/StorageCommitmentScp/Plugin.cpp#l38), the `Lookup` callback looks like the following:

```c
virtual OrthancPluginStorageCommitmentFailureReason Lookup(const std::string& sopClassUid, const std::string& sopInstanceUid)
```

which doesn't include the argument `handler` since the function itself is a member function of `handler`. We could simply access `handler` by the `this` keyword. As for writing Python plugins, the `Lookup` callback looks like the following:

```py
def StorageCommitmentLookup(sopClassUid, sopInstanceUid, data):
```

If the order of the arguments is changed to `[handler, sopClassUid, sopInstanceUid]` (as you mentioned), the `Lookup` callback in Python would look like:

```py
def StorageCommitmentLookup(data, sopClassUid, sopInstanceUid):
```

which adds a new argument before `sopClassUid` when compared to C++.

Reply:
You're perfectly right !  Let's keep it as is.